system_keyspace: remove flushes when writing to system tables

gusev-p opened the pull request:
There are several system tables with strict durability requirements. This means that if we have written to such a table, we want to be sure that the write won't be lost in case of node failure. We currently accomplish this by accompanying each write to these tables with `db.flush()` on all shards. This is 
expensive, since it causes all the memtables to be written to sstables, which causes a lot of disk writes. This overheads can become painful during node startup, when we write the current boot state to `system.local`/`system.scylla_local` or during topology change, when `update_peer_info`/`update_tokens` write to `system.peers`.

In this series we remove flushes on writes to the `system.local`, `system.peers`, `system.scylla_local` and `system.cdc_local` tables and start using schema commitlog for durability.

Fixes: #15133

gusev-p commented:
There are a few questions left to resolve:
  * `enable_features_on_startup` reads from `system.scylla_local` and `system.topology`, we guarantee that features in `system.scylla_local` will be available before schema commitlog replay, since we left flush in `system_keyspace::save_local_enabled_features`, but we don't have the same guarantee for the `system.topology` table. This problem existed before this series.
  * we need to decide what to do with the truncation records check in `initialize_system_keyspace`. This check was outdated prior to this series, as it only included schema tables and did not take into account other tables using schema commitlog.
  * do we need other system tables to be switched to schema commitlog for consistency?
  * asses the performance implications of this series. We could make the things worse since every write to the table using schema commitlog incurs a logical `fsync` (`db::commitlog::sync_mode::BATCH`, `use_o_dsync` schema commitlog options).

piodul commented:
> > enable_features_on_startup reads from system.scylla_local and system.topology, we guarantee that features in system.scylla_local will be available before schema commitlog replay, since we left flush in system_keyspace::save_local_enabled_features, but we don't have the same guarantee for the system.topology table. This problem existed before this series.
> 
> cc @piodul

The issue was fixed in a followup, in this commit: https://github.com/scylladb/scylladb/pull/15168/commits/371b6403091069108e92dccdfc15c2389d917fb1 . The `system.topology` table is explicitly flushed after we modify `enabled_features`/`supported_features` columns, and also on snapshot transfer.

piodul commented on db/system_keyspace.cc:
```diff
@@ -2341,6 +2331,9 @@ future<std::set<sstring>> system_keyspace::load_local_enabled_features() {
 future<> system_keyspace::save_local_enabled_features(std::set<sstring> features) {
     auto features_str = fmt::to_string(fmt::join(features, ","));
     co_await set_scylla_local_param(gms::feature_service::ENABLED_FEATURES_KEY, features_str);
+    // Flush the table so that the value is available on boot before commitlog replay.
+    // database::maybe_init_schema_commitlog() depends on it.
+    co_await force_blocking_flush(SCYLLA_LOCAL);
 }
```
The `save_local_enabled_features` function is called in two places: in `persistent_feature_enabler::enable_features()`, but also in `storage_service::topology_state_load()`.

In the first case, the flush is necessary because it is related to the legacy features-in-gossip path: we call it after enabling features based on gossip, and we use the features saved in `system.scylla_local` for the feature check in startup.

The second case is related to features-in-raft; we update `system.scylla_local` based on the group 0 state only for legacy purposes, like with `system.peers` etc. In that case you will most certainly want to avoid flushing as `system.scylla_local` is no longer a source of truth for the features and we don't use it in the feature check on startup.

gusev-p replied:
done, thanks

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3469/

kostja commented on main.cc:
```diff
@@ -544,6 +544,85 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
+    // Needs to happen before replaying the schema commitlog, which interprets
+    // replay position in the truncation record.
+    // Needs to happen before system_keyspace::setup(), which reads truncation records.
+    db.local().get_tables_metadata().for_each_table([] (table_id, lw_shared_ptr<replica::table> table_ptr) {
```
I think this code should be removed or hidden before an upgrade switch. It should be possible to truncate system tables.

kostja replied:
@tgrabiec

tgrabiec replied:
Schema tables cannot be safely truncated at run-time, because they must match in-memory state. If you want to do that, you have to stop the node and delete the sstables.

kostja replied:
It is not true for just any schema table @tgrabiec 
I think people who truncate schema tables at runtime know what they are doing.

kostja replied:
Any meddling with schema tables by a user is not allowed. By not allowing ourselves truncate we just shoot ourselves in the foot.

tgrabiec replied:
You mean it should be allowed internally, but disabled for the user?

What's the use case?

Anyhow, the system is really not prepared for supporting that. The node will crash afterwards. So if we want to support that, it takes much more work than just removing this check.

tgrabiec replied:
The proper way to achieve clean state of schema tables is to issue DROP statements.

kostja replied:
Can you come up with a case when e.g. view build status or column mapping history tables will get broken if truncated? These are not "schema" tables, perhaps, but system tables. I don't have a use case. I can't imagine a user trying to truncate system tables/schema tables, but now discussing it with you perhaps I think it is a bad idea even for our own staff. However, I think the guardrails for this are redundant.

tgrabiec replied:
The guardrails we have are specifically for system_schema.* tables. Other system tables may be fine to truncate.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -72,14 +72,9 @@ namespace {
             system_keyspace::TOPOLOGY,
-            system_keyspace::CDC_GENERATIONS_V3,
-            system_keyspace::TABLETS,
+            system_keyspace::CDC_GENERATIONS_V3
```
nit: we usually leave ',' after the last element to make patches extending the table less noisy.

gusev-p replied:
fixed

kbr-scylla commented on replica/database.hh:
```diff
@@ -445,6 +444,7 @@ private:
 
     // Provided by the database that owns this commitlog
     db::commitlog* _commitlog;
+    bool _readonly;
```
Please comment on this flag and its relationship with commitlog

gleb-cloudius replied:
Can we actually use `_commitlog == nullptr" as readonly indication?

kbr-scylla commented on replica/table.cc:
```diff
@@ -2040,6 +2041,24 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
+    if (_virtual_writer) {
+        return nullptr;
+    }
+    if (_readonly) {
```
Let's add ` [[unlikely]]` to both `if`s.

gusev-p replied:
done.

BTW, do you know the practical consequences of this? Does it really help with mispredictions?

kbr-scylla commented on main.cc:
```diff
@@ -569,22 +569,22 @@ static void initialize_system_keyspace(sharded<db::system_keyspace>& sys_ks,
     // This should be done before commitlog starts replaying
     // since some features affect storage.
+    // We can use system.scylla_local table before replaying commitlog since
+    // we flush memtables when we write feature.
```
nit: feature**s**

kbr-scylla commented on replica/database.cc:
```diff
@@ -1000,6 +1000,13 @@ future<> database::create_local_system_table(
     co_await add_column_family(ks, table, std::move(cfg));
 }
 
+void database::mark_table_ready_for_writes(const schema_ptr& schema) {
```
Is this function idempotent? If not, maybe we should do a sanity check that this wasn't done before

gusev-p replied:
ok

kbr-scylla commented:
> we need to decide what to do with the truncation records check in initialize_system_keyspace. This check was outdated prior to this series, as it only included schema tables and did not take into account other tables using schema commitlog.

Please create an issue (we should fix the check). It can be done as a follow-up to this PR.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -82,6 +82,7 @@ namespace {
             system_keyspace::GROUP0_HISTORY,
```
Schema commitlog is more durable then regular one because it is always in `db::commitlog::sync_mode::BATCH` mode. In this mode writes wait for data to actually be on disk at least in commitlog before returning.

kbr-scylla replied:
Ok, but given that the tables we're dealing with were using wait_for_sync_to_commitlog -- we didn't gain any durability, right?

gleb-cloudius replied:
If this is the case then right as far as I see.

kostja replied:
Yes, in fact that's what I do on my branch which sparked this pull request. Please feel free to add to this series or do a follow up.

gusev-p replied:
> I don't see why it should be different for schema commit log.

`table::flush` does `fsync` on the just written sstables (`scylladb/sstables/mx/writer.cc/writer::close_data_writer`), writing to schema commitlog without 
`use_o_dsync` won't `fsync`, the data can be lost in case of power outage, so this commit worsens durability.

avikivity replied:
But does this schenario (write to schema commitlog without use_o_dsync) ever happen? apart from developer/test settings.

tgrabiec replied:
even without o_dsync we do sync, batch_cycle() will call sync(), which calls _file.flush(), which calls fdatasync().  

```

future<>
posix_file_impl::flush() noexcept {
    if ((_open_flags & open_flags::dsync) != open_flags{}) {
        return make_ready_future<>();
    }
    return engine().fdatasync(_fd);
}

```

gusev-p replied:
> even without o_dsync we do sync, batch_cycle() will call sync(), which calls _file.flush(), which calls fdatasync().

I didn't notice that, thank you. There is no problem then, I suppose.

gusev-p replied:
> But does this schenario (write to schema commitlog without use_o_dsync) ever happen? apart from developer/test settings.

`use_o_dsync` is accessible through `config.cc`, so the use could set it to `false`.

xemul commented on db/system_keyspace.cc:
```diff
@@ -1438,14 +1438,21 @@ future<> system_keyspace::build_bootstrap_info() {
+
+    // It's safe to load truncation records here, since
+    // all the writes to the table are followed by flushes.
+    co_await cache_truncation_record();
```
Just pull this out of system_keyspace.cc and put in the needed place on main(), please

gusev-p replied:
> Just pull this out of system_keyspace.cc and put in the needed place on main(), please

Yes, done.

xemul commented on db/system_keyspace.cc:
```diff
@@ -1438,14 +1438,21 @@ future<> system_keyspace::build_bootstrap_info() {
+    co_await cache_truncation_record();
+}
+
+future<> system_keyspace::setup_after_commitlog() {
```
And please leave the .setup() intact, it's going to be removed anyway

gusev-p replied:
I inlined it into the proper places in main.

xemul commented on main.cc:
```diff
@@ -524,6 +524,26 @@ sharded<locator::snitch_ptr> *the_snitch;
 sharded<service::storage_proxy> *the_storage_proxy;
 }
 
+static void initialize_local_info(sharded<db::system_keyspace>& sys_ks,
```
Please, write a patch comment describing what's the point of this change. Are you going to re-use this function in one of the next patches?

gusev-p replied:
>> Are you going to re-use this function in one of the next patches?

No, the giant `scylla_main` just makes me scared and I can't reason effectively about the code in that form. So it's just a readability refactoring.

xemul commented on tools/schema_loader.cc:
```diff
@@ -339,7 +339,7 @@ struct sstable_manager_service {
         : feature_service(gms::feature_config_from_db_config(dbcfg))
         , dir_sem(1)
-        , sst_man(large_data_handler, dbcfg, feature_service, tracker, memory::stats().total_memory(), dir_sem) {
+        , sst_man(large_data_handler, dbcfg, feature_service, tracker, memory::stats().total_memory(), dir_sem, []{ return locator::host_id{}; }) {
```
Note to ourselves that schema_loader.cc already has "artificial token metadata"

xemul commented on tools/scylla-sstable.cc:
```diff
@@ -2876,9 +2876,10 @@ Validate the specified sstables:
         auto& dbcfg = *pdbcfg;
         gms::feature_service feature_service(gms::feature_config_from_db_config(dbcfg));
         cache_tracker tracker;
-        dbcfg.host_id = locator::host_id::create_random_id();
```
The scylla-sstable doesn't have token metadata indeed, but the cost of having it as large as

```
    sharded<locator::shared_token_metadata> token_metadata;
    token_metadata.start([] () noexcept { return db::schema_tables::hold_merge_lock(); }, locator::token_metadata::config{}).get();
    auto stop_token_metadata = deferred_stop(token_metadata);
```

I would try to avoid the ugliness with keeping the resolve_host_id() noncopyable_function on sstables manager and put the explicit manager->tokens dependency.

Longer term -- it's probably worth making topology _not_ be the part of token_metadata, in that case service that just need topology (e.g. -- messaging service does need it already) won't have to reference token metadata. @kbr-scylla , WDYT?

kbr-scylla replied:
> Longer term -- it's probably worth making topology not be the part of token_metadata, in that case service that just need topology (e.g. -- messaging service does need it already) won't have to reference token metadata. @kbr-scylla , WDYT?

Yes

avikivity replied:
I think using a function is reasonable here.

nyh commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3500/

Please look at the gazillion failed tests, I think it's related to your patch.

Just one random example is https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3500/testReport/junit/(root)/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_describe_release_1/ - apparently `system.schema_aggregates` is missing in the return value of DESC results but present in system_schema. I don't know how this is possible.

gusev-p commented:
v2:
  * fixed `update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_garbage_members_from_group0_after_abort_decommission`; virtual tables missed from system tables because of the broken load dependencies, we need to call `db::schema_tables::save_system_schema` after `db::initialize_virtual_tables` and `db::legacy_schema_migrator::migrate`. The fix required some refactorings, in particular, [we inlined](https://github.com/scylladb/scylladb/pull/15279/commits/ba0df51faf11e62e7eb94763fa5a3317bf1a7d8b) `system_keyspace::setup`.

[all the changes combined](https://github.com/scylladb/scylladb/compare/dc1e0636374b9a04b2ee181c5a14a502bc998703..8ab40b078f4ffd75c87dd43e0448e92264357412): 

new/changed commits: 
* [system_keyspace: move initialize_virtual_tables into virtual_tables.hh](https://github.com/scylladb/scylladb/pull/15279/commits/2e08ec7b615d3ed2ee5cb143b0a1c39a637dfc17)
* [system_keyspace: refactor save_system_schema function](https://github.com/scylladb/scylladb/pull/15279/commits/4ab1169592e724f8e3d1125bfccb97618c4c70e7)
* [main.cc: inline and split system_keyspace.setup](https://github.com/scylladb/scylladb/pull/15279/commits/ba0df51faf11e62e7eb94763fa5a3317bf1a7d8b)
* [main.cc: move schema commitlog replay earlier](https://github.com/scylladb/scylladb/pull/15279/commits/f949d4be856b6e4740d7434fd76a69f6d27fef33)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3541/

gusev-p commented:
v3:
  * [small changes](https://github.com/scylladb/scylladb/compare/8ab40b078f4ffd75c87dd43e0448e92264357412..2702858432bee0f1694e88ee6c4f70b12a58e354) to fix review comments

gusev-p commented:
v4:
  * fix [sstables_format_selector::start](https://github.com/scylladb/scylladb/compare/2702858432bee0f1694e88ee6c4f70b12a58e354..cab889a7ae8979faf397b17c39603659466fa818)

xemul commented on main.cc:
```diff
@@ -524,6 +524,26 @@ sharded<locator::snitch_ptr> *the_snitch;
 sharded<service::storage_proxy> *the_storage_proxy;
 }
 
+static void initialize_local_info(sharded<db::system_keyspace>& sys_ks,
```
Please, note, that this function implicitly "knows" that it's run in async context. This is a time-bomb, because ~~when~~ if we do anything with the way it's called it will take extra eye+brain cycles to understand it. It's worth either making it a coroutine right at once, or at least add a comment that it's run in async context. Can be done as a followup if this PR is going to get merged the way it is now

gusev-p replied:
Good point, no problem to turn it into coroutine in this PR, I'll do this.

tgrabiec replied:
Initialization is unlikely to be coroutinized as a whole. It's better to use seastar threads for that because it's easier to debug. With threads it's possible to obtain full backtrace, and we can see where the main thread is blocked.

avikivity replied:
Agree. It's enough to append _thread to the name to clearly mark it.

gusev-p replied:
ok, will add `_thread` suffix

xemul commented on sstables/sstables_manager.cc:
```diff
@@ -98,8 +99,8 @@ storage_manager::config_updater::config_updater(const db::config& cfg, storage_m
-const locator::host_id& sstables_manager::get_local_host_id() const {
-    return _db_config.host_id;
+locator::host_id sstables_manager::get_local_host_id() const {
+    return _resolve_host_id();
```
I don't want to stand in the way of progress, but I still don't like it

gusev-p replied:
> Please add a comment somewhere what the precondition for calling get_local_host_id is.

done

xemul commented on db/system_keyspace.hh:
```diff
@@ -285,7 +285,6 @@ public:
     future<> make(
             locator::effective_replication_map_factory&,
             replica::database&,
-            db::config&,
```
I know how it happened that this arg became unused.
<boring mode>
The db::config is the generic scylla-wide configuration container that used to sit on a database object from day-0. And everybody got used to the fact that they can get this config out of database. Eventually, the intention became that every service should have its own specialized config, e.g. database got its `database_config` thing, so did many other services. The db::config was supposed to have its scope limited to main.cc (and some of its pals) and be used to generate services' configs. Here in system_keyspace::make() the db::config was sort of used like that, but at some point the database& argument appeared and instead of db::config& cfg the db.get_config() took its place.
</boring mode>

xemul commented on main.cc:
```diff
@@ -1425,6 +1422,9 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                 }
             }
 
+            sys_ks.local().build_bootstrap_info().get();
```
nit: Can happen as late as the `group0_service.start().get();` happens

gusev-p replied:
The idea behind this function (`initialize_system_keyspace`) is the opposite - load things eagerly as early as possible so that we don't have to worry about load order later. Otherwise it's difficult to reason about what should be loaded before what, and comments doesn't help much here.

xemul replied:
If the function does "load things eagerly as early as possible so that ..." then it shouldn't be named like that

gusev-p replied:
What do you suggest?

xemul replied:
a) Not to introduce `initialize_system_keyspace()` at all
b) (optionally) put call for `sys_ks.local().build_bootstrap_info()` lower in the main() codeflow

xemul commented on db/sstables-format-selector.cc:
```diff
@@ -53,8 +53,11 @@ future<> sstables_format_selector::maybe_select_format(sstables::sstable_version
-    _features.local().me_sstable.when_enabled(_me_feature_listener);
-    _features.local().md_sstable.when_enabled(_md_feature_listener);
+    // The listener may fire immediately, create a thread for that case.
+    co_await seastar::async([this] {
```
This is a bug-fix, isn't it?

gusev-p replied:
Probably, but before my movements in `scylla_main` it didn't matter - features were loaded after
`sstables-format-selector::start`.

tgrabiec replied:
This wrapping belongs more to when_enabled() impl.

gusev-p replied:
> This wrapping belongs more to when_enabled() impl.

Agree, but I already did a lot of refactorings in this PR, let's leave that for another.

tgrabiec replied:
> let's leave that for another.

Ok

avikivity replied:
We should make it explicit that listeners can yield (make them return a future). Threads are good in some local scope, but when gluing different components, we need to make the possibility of yielding clear.

gusev-p replied:
I [created an issue](https://github.com/scylladb/scylladb/issues/15347), will fix it in a follow-up.

xemul commented on main.cc:
```diff
@@ -545,6 +545,84 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
+        sharded<gms::feature_service>& feature_service)
+{
+    supervisor::notify("starting system keyspace");
+    sys_ks.start(std::ref(qp), std::ref(db), std::ref(snitch)).get();
```
No, no, any `sharded<service>.start(args...)` call should remain in main(). The ultimate goal (that's almost reached in cql_test_env's do_with()) is to make main look like

```
    sharded<service> srv;
    co_await srv.start(deps...);
    auto stop_srv = defer([&srv] { co_await srv.stop(); });
    co_await srv.invoke_on_all(&service::start);

    ... // same sequence for other services
```

all the other stuff, like dist_loader::init_sys_ks(), maybe_init_schema_commitlog(), etc you can scatter the weay you like

gusev-p replied:
why?

tgrabiec replied:
presumably because of RAII for stopping.

xemul replied:
Yes, because of `auto stop = defer([&srv] { co_await srv.stop(); });` thing. Thank you, Tomek

xemul commented on main.cc:
```diff
@@ -545,6 +545,84 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
+    // replay position in the truncation record.
+    db.local().get_tables_metadata().for_each_table([] (table_id, lw_shared_ptr<replica::table> table_ptr) {
+        if (table_ptr->schema()->ks_name() == db::schema_tables::NAME) {
+            if (table_ptr->get_truncation_record() != db_clock::time_point::min()) {
```
A note (for followup, I suppose). The truncation records had been populated one step above. There's no need to walk the whole set of tables metadata manually filtering out schema_tables::NAME ones. We can patch the truncation records loader so that _it_ catches the schema_tables::NAME updates

gusev-p replied:
We need to rethink, whether we need this code at all, because if we follow this route we need to include in the `if` all tables that use schema commitlog, and this is not feasible, we want to retain truncates for them.

tgrabiec replied:
> we want to retain truncates for them.

For which tables?

tgrabiec replied:
Note, we also have this in database::truncate_table_on_all_shards():
```
    if (s->ks_name() == db::schema_tables::NAME) {
        throw std::runtime_error(format("Truncating of {}.{} is not allowed.", s->ks_name(), s->cf_name()));
    }
```

xemul commented on main.cc:
```diff
@@ -545,6 +545,84 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
+    // Re-enable previously enabled features on node startup.
+    // This should be done before commitlog starts replaying
+    // since some features affect storage.
+    feature_service.local().enable_features_on_startup(sys_ks.local(), raft_topology_change_enabled(cfg)).get();
```
Reeeeally? The initialize_system_keyspace() is supposed to initialize system keyspace. Well, maybe init schema commitlog :P (below). Enabling features looks completely alien here

gusev-p replied:
What can we do? Don't extract `initialize_system_keyspace`? It's like if I don't see the problem maybe it'll go away. We can bury this dependency between features and `system_keyspace` boot load in the sea of `scylla_main` and nobody will notice it until the next code movements which will break it.

tgrabiec replied:
But feature_service is not system keyspace, it depends on it. Everything we do later in main depends on system keyspace, but you don't put it in  initialize_system_keyspace().

xemul replied:
> Don't extract initialize_system_keyspace? 

Yes. The whole main() really consists of 3 logical blocks
- configure seastar app
- start sharded services
- do necessary initialization steps
- ... then call join_cluster()

All three are intermixed :( The 3rd is the largest. What this patch does is peeks random pieces from it and puts them into a separate function (of a strange name). Not that I consider this code to be at good shape, but this particular change doesn't make it better

gusev-p replied:
> But feature_service is not system keyspace, it depends on it. Everything we do later in main depends on system keyspace, but you don't put it in initialize_system_keyspace().

1. We put into `initialize_system_keyspace` things related to `system_keyspace` initialization;
2. We put `enable_features_on_startup` in it because we have two way dependency, `system_keyspace` depends on features and features depend on `system_keyspace`;
3. This circular dependency exists regardless of whether we extract `initialize_system_keyspace` or not. The separate function just highlights the problem.

tgrabiec replied:
I see.

gusev-p replied:
> All three are intermixed :( The 3rd is the largest. 

ok, then why not split it into functions to make the code manageable, like we do in all other places, where a single function grows too big? I think the code should be like this:
```
sharded<Service1> service1;
sharded<Service2> service2;
...
void initialize_service1() {
  service1.start(...);
  // other necessary load//boot/wiring stuff
}

void initialize_service2() {
...
}

void initialize_services() {
  initialize_service1();
  initialize_service2();
  ...
}
```
We can put shutdown actions in a container and run them in the appropriate place, e.g. at the end of `scylla_main`. We can even create a simple factory helper
```
template<T, Args...>
void init_with_stop(sharded<T>& service, Args... args) {
  service.start(args...);
  if constexpr(T has stop method) {
    shutdown_actions.push_back([]() { service.stop(); });
  }
}
```
and get rid of ~90% of `defer_verbose_shutdown` boilerplate

But this can't be done in one go, the code is too convoluted, that's why I 'peek random pieces from it and puts them into a separate function'. I don't know, what's strange about the name 'initialize' though.

avikivity replied:
@gusev-p your suggestion will break the rollback ordering between services called with init_with_stop() and random objects on the frame. We can fix it by making init_with_stop() return an `undo` object.

```c++
undo stop_blah = init_with_stop(blah, arg1, arg2);
```

avikivity replied:
Here are some ways to solve the feature_service <-> system_keyspace dependency loop

1. store features outside the database (BAD)
2. read features from sstables + mutation_reader directly
3. create a system_keyspace object with instructions to only read the features table, construct a private database + query_processor on top, read features, destroy these temporary objects, initialize features with result of previous read, continue (maximum reuse but a long road to get there)

This problem is also seen in filesystems which store their own metadata in files.

gusev-p replied:
> your suggestion will break the rollback ordering between services called with init_with_stop() and random objects on the frame. We can fix it by making init_with_stop() return an undo object.

If we had simple initialisation code structure like 
```
void initialize_services() {
  initialize_service1();
  initialize_service2();
  ...
}
```
we wouldn't have random objects on the frame.

This doesn't preclude extracting the `initialize_` functions though, we can return the destroy object from it and store it on the stack.

xemul replied:
> ok, then why not split it into functions to make the code manageable, like we do in all other places

I read this question as "why don't we make main being structured already". I don't know how to answer it, sorry :) I can only rant about this place was much worse in this sense 4 years ago and there's still some route to be walked with it. You're welcome to join, but from my POV _this_ particular change doesn't make it better.

> I don't know, what's strange about the name 'initialize' though.

Because this brand new `initialize_system_keyspace()` does much more than what it name says. I disagree that enabling features on startup is the part of system keyspace initialization. The fact that system keyspace cannot be initialized without enabled features still doesn't justify it. Some other systems also need enabled features and they thus piggy-back system keyspace initialization thus not contributing to "structured and manageable code" concept.

xemul replied:
FYI, there's another ... bag of accumulated not-immediately-obviously-structured (but apparently mostly working) code out there. Check the `storage_service::join_cluster()` -> `storage_service::join_token_ring()` (-[partially]-> `storage_service::bootstrap()`) methods.

xemul replied:
> I don't know, what's strange about the name 'initialize' though.

Maybe we can approach this from slightly different angle. If it's indeed the initialization of the system keyspace, then try making this code be a (preferably non-static) method of the system_keyspace class and place it in db/system_keyspace.cc. _That_ would make it obvious if the change makes the code more structured or less structured.

xemul replied:
> I mean, why not start splitting scylla_main into functions right now?

I'm all for it, but the `initialize_system_keyspace` doesn't fit. As I wrote somewhere else in this PR, turning this function into a system_keyspace's method -- likely does.

> There are two points in this discussion:
>   *  should we extract initialize_system_keyspace_thread or not;

Provided this function is system_keyspace's method -- yes, we should try to. Provided it's not -- no we shouldn't

> I can make it an instance method of system_keyspace, but it doesn't feel right to me. I'm more for some scylla_configurator class, which contains all the wiring logic and can be reused in boost tests, for example.

Yes, we have a related issue: #2795. If the configurator you're talking about in question would suite both -- main() and single_node_cql_test_env::do_with() -- I'll likely change my mind about `initialize_system_keyspace`

xemul replied:
> If we had simple initialisation code structure like
> 
>     void initialize_services() {
>       initialize_service1();
>       initialize_service2();
>       ...
>     }

From what I've seen so far, that's very likely utopian goal. We can definitely try to arrange `sharded<service>::start`-s and `invoke_on_all(&service::start)`-s in a plain sequence, but some initialization parts will still be intermixed between different services. Recent invasion of group0_<whatever> (and related) in main proves that. Three ways I can see how to overcome the latter.

1. untie it with either inter-dependent semaphores/condvars/on_...-notifications (like you did here).
2. split services into smaller parts introducing more sharded<Foo> stuff in main
3. relax and make the "bottom-half" part of the initialization explicitly intermixed

gusev-p replied:
> Yes, we have a related issue: https://github.com/scylladb/scylladb/issues/2795. If the configurator you're talking about in question would suite both -- main() and single_node_cql_test_env::do_with() -- I'll likely change my mind about initialize_system_keyspace

I can create a new class in a separate file, `scylla_service_host` (sounds more appropriate), move my new functions into it, along with the services they care about, reuse this class in `scylla_main` and `cql_test_env.cc`. This will set the direction for the refactoring. It is impossible to do it from start to finish in one PR.

gusev-p replied:
Do I understand your suggestion right:
  * revert my changes about extracting functions from `scylla_main` in this PR;
  * try to do the same in another PR, but with your implication that these changes will not be successful and will most likely be reverted too. ("you'll likely see that this brand new scylla_service_host thing is just becoming that very main() renamed and moved to another file.")

This doesn't work on a motivational level. I can only revert the changes and never touch this topic again.

xemul replied:
> revert my changes about extracting functions from scylla_main in this PR;

Not quite. I accept the `initialize_local_info_thread` change and only suggest not to introduce `initialize_system_keyspace` _here_, because this change grows out of "remove flushes when writing to system tables" scope. As  you told above: "It is impossible to do it from start to finish in one PR."

> try to do the same in another PR, but with your implication that these changes will not be successful and will most likely be reverted too. ("you'll likely see that this brand new scylla_service_host thing is just becoming that very main() renamed and moved to another file.")

I didn't say it would not be successful. I just shared my thoughts that structuring main is more than moving its pieces into functions with "descriptive" names; and my expectation that this code move would end up being just code move.

> This doesn't work on a motivational level. I can only revert the changes and never touch this topic again.

I feel your pain, some of my changes were rejected by maintainers as well. Apparently every time a change is submitted, it can be asked to be reworked or even thrown away. There's even the explicit "request changes" option in the "submit my review" form here on github.

gusev-p replied:
> only suggest not to introduce initialize_system_keyspace here, because this change grows out of "remove flushes when writing to system tables" scope. As you told above: "It is impossible to do it from start to finish in one PR."

The essence of the change was not extracting `initialize_system_keyspace` by itself, but all the preparatory work - gather its pieces from various parts of `scylla_main` and making them work together - splitting `sstables_format_selector`, introducing `on_system_tables_loaded` interface, inlining `system_keyspace.setup` parts into the correct places. Those changes were dictated by their own reasons, but in the end we got something relatively cohesive, and now we are just refusing to put a name on it for some reason.

If we want to structure `scylla_main`, then the work above is what we have to do to many other services - make their initialisation routines as independent from each other as possible and define interfaces between them where necessary. I agree with you that this is more than 'moving pieces and giving names'.

xemul replied:
> The essence of the change was not extracting `initialize_system_keyspace` by itself, but all the preparatory work - gather its pieces from various parts of `scylla_main` and making them work together - splitting `sstables_format_selector`, introducing `on_system_tables_loaded` interface, inlining `system_keyspace.setup` parts into the correct places. Those changes were dictated by their own reasons, but in the end we got something relatively cohesive, and now we are just refusing to put a name on it for some reason.

So let's merge those preparatory, gathering, making-them-work-together, splitting, introducing, etc. changes! Just keep them in one bundle inside main.cc _or_ place in the system_keyspace::initialize_on_boot_or_whatever() non-static method to show that it's indeed self-contained. Moving those into main-local static method is code move for the sake of the code move itself. The amount of arguments you need to carry along and templatized list of notify-me services speaks for itself from my perspective.

> If we want to structure `scylla_main`, then the work above is what we have to do to many other services - make their initialisation routines as independent from each other as possible and define interfaces between them where necessary. I agree with you that this is more than 'moving pieces and giving names'.

:+1:

gusev-p replied:
ok, I'll leave it 'as is' inside `scylla_main`, moving it into `system_keyspace` just creates too many strange dependencies to it (e.g. `supervisor::notify`)

gusev-p replied:
But I really don't get the arguments here, they sound arbitrary to me.

I think the code of the form
```
initialize_service1();
initialize_service2();
...
initialize_service158();
```
is better than then same code where all the functions are inlined. At least we can use function names instead of line numbers when we reason about the code.

xemul replied:
It's better, but this patch doesn't make it so. The change is

```patch
@@ -10,9 +10,7
     mess_a
     mess_b
     mess_c
-    mess_d
-    mess_e
-    mess_f
+    initialize_foo()
     mess_g
     mess_h
     mess_i
@ -20,6 +20,11
  
  
  
+void initialize_foo() {
+    mess_d
+    mess_e
+    mess_f
+}
  
  
  
```

you say: "it's a step in right direction"
I say: "I doubt it, but let's better try to make at least two or, better, three such steps in a separate PR"

gusev-p replied:
ok

gusev-p commented:
v5:
 * [materialised views dtests fix](https://github.com/scylladb/scylladb/pull/15279/commits/0935330cb096705730ea9e813c848775278d890e), required some other [small changes](https://github.com/scylladb/scylladb/compare/cab889a7ae8979faf397b17c39603659466fa818..a6949046818445effabe5fa48204f291412e69e6).

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -1526,15 +1526,11 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
         // That way if a view seeks information about its base table it's guarantied to find it.
         co_await max_concurrent_for_each(tables_diff.created, max_concurrent, [&] (global_schema_ptr& gs) -> future<> {
-            co_await db.add_column_family_and_make_directory(gs);
+            co_await db.add_column_family_and_make_directory(gs, false);
```
This patch doesn't seem to be a full solution to the problem:

> 
> This problem was demonstrated by materialised view dtests.
> A view is created on all nodes. On some nodes it will be created
> earlier than on others and the view rebuild process will start
> writing data to that view on other nodes, where mark_table_as_writable
> has not yet been called.

The table may have not yet been created on "other" node, and the update will fail on missing table. So it doesn't help that we mark it as ready for writes sooner, it may still not be soon enough. Another thing is that we shouldn't hit this problem because what should happen in this case is that such update should trigger schema pull on the replica due to an unsynced schema version and should proceed after schema in synced, which means that the table should be ready on all the shards by the time the update is attempted.

Maybe you hit https://github.com/scylladb/scylladb/issues/15048

kostja replied:
@tgrabiec re o_dsync, I don't see any reason to not use it if we know we want all writes to the file to be done in sync mode. 

The reason it is not sufficient to provide duarbility is that not all file systems support o_dsync, on these you need to additionally flush the write explicitly. But we don't run the commit log on such file systems. In all practical installations it is going to be enabled and give a significant performance boost.
It should always be set for schema commit log.

gusev-p replied:
> he table may have not yet been created on "other" node, and the update will fail on missing table. So it doesn't help that we mark it as ready for writes sooner, it may still not be soon enough. Another thing is that we shouldn't hit this problem because what should happen in this case is that such update should trigger schema pull on the replica due to an unsynced schema version and should proceed after schema in synced, which means that the table should be ready on all the shards by the time the update is attempted.

My reasoning around this approach:
* I managed to reproduce the views problem steadily on  the dtest `update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_garbage_members_from_group0_after_abort_decommission[leaving_token_ring-]`, this test failed consistently on my dev machine;
* The failures happen only with my commit where I trigger an internal_error when the `table.commitlog()` is accessed on the table, which has not yet been made writable through `table::mark_ready_for_writes`;
* In the failure backtrace I saw "Failure to apply mutations from <some ip>", where some ip != this node ip;
* No failures if we apply this commit;
* Looking at the `view_builder` code I saw a lot of `catch(no_such_column_family)` e.g. in `view_builder::on_create_view`, so I though it was expected when we build the view on another node and the current what doesn't have the view table yet, maybe rebuild will be retried later.

tgrabiec replied:
Ok, I guess to avoid regressions we have to make this patch. Probably, the reason why it didn't fail before was that we mark the schema as synced when we create the view in add_column_family():
```
    schema->registry_entry()->mark_synced();
``` 
which is wrong (https://github.com/scylladb/scylladb/issues/14146) and we should move this mark out of there and do it after all shards have the table. Then this patch can be reverted.

avikivity replied:
> @tgrabiec re o_dsync, I don't see any reason to not use it if we know we want all writes to the file to be done in sync mode.
> 
> The reason it is not sufficient to provide duarbility is that not all file systems support o_dsync, on these you need to additionally flush the write explicitly. But we don't run the commit log on such file systems. In all practical installations it is going to be enabled and give a significant performance boost. It should always be set for schema commit log.

Just curious, which file systems don't support O_DSYNC? I thought it was implemented purely in VFS.

kostja replied:
https://twitter.com/AlexMillerDB/status/1532280950779891712 recent relevant discussion about O_DSYNC on Darwin

kbr-scylla replied:
Reason not to use O_DSYNC in tests: https://github.com/scylladb/scylladb/pull/12542

kbr-scylla replied:
Perhaps a way out would be to enable O_DSYNC for schema commitlog but disable segment preallocation: https://github.com/scylladb/scylladb/issues/15349

kostja replied:
I think Scylla is on a safe side since it uses O_DIRECT. For buffered writes, things can go wrong in many other ways. This PostgreSQL page summarizes the issue: https://wiki.postgresql.org/wiki/Fsync_Errors

kostja replied:
In any case we don't do fsync-failure-testing as is suggested here, https://www.usenix.org/system/files/atc20-rebello.pdf

avikivity replied:
What's wrong with disabling O_DSYNC for tests? they don't need crash recovery.

avikivity replied:
And we don't care about Darwin (the OS).

avikivity replied:
I expect that NFS emulates O_DSYNC by sending flushes after each write.

kbr-scylla replied:
> What's wrong with disabling O_DSYNC for tests? they don't need crash recovery.

Perhaps we'd like to test crash recovery. Especially with these changes around durability.

tgrabiec commented on db/system_keyspace.cc:
```diff
@@ -2302,9 +2297,14 @@ future<std::set<sstring>> system_keyspace::load_local_enabled_features() {
+    // Flush the table so that the value is available on boot before commitlog replay.
+    // database::maybe_init_schema_commitlog() depends on it.
+    if (durable) {
+        co_await force_blocking_flush(SCYLLA_LOCAL);
```
The name "durable" is not accurate, since the write is durable even if this parameter is false, it's just that it won't be visible on boot before replay. Maybe "visible_before_cl_replay"?

gusev-p replied:
Right. Maybe just `bool flush` + comment?

avikivity replied:
Better to have a long an unambiguous name.

gusev-p replied:
ok

tgrabiec commented on db/system_keyspace.cc:
```diff
@@ -2302,9 +2297,14 @@ future<std::set<sstring>> system_keyspace::load_local_enabled_features() {
 }
 
-future<> system_keyspace::save_local_enabled_features(std::set<sstring> features) {
+future<> system_keyspace::save_local_enabled_features(std::set<sstring> features, bool durable) {
```
What about SSTABLE_FORMAT_PARAM_NAME? Don't we need it early too, so that we don't write sstables in old format during cl replay?

gusev-p replied:
yes, fixed

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3558/

gusev-p commented on main.cc:
```diff
@@ -1378,7 +1378,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // 2. cql_test_env() doesn't do it
             // 3. need to check if it depends on any of the above steps
-            sys_ks.local().setup(messaging).get();
+            sys_ks.local().setup_before_commitlog().get();
```
In `initialize_system_keyspace` we now have
```
"starting system keyspace"
"loading system sstables"
"starting schema commit log"

```
What else `supervisor::notify` do you suggest?

gusev-p commented on main.cc:
```diff
@@ -545,6 +545,84 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
+        }
+    });
+
+    auto sch_cl = db.local().schema_commitlog();
```
hmm, probably yes, I'll think about it, thanks

gusev-p replied:
fixed

avikivity commented:
> * asses the performance implications of this series. We could make the things worse since every write to the table using schema commitlog incurs a logical `fsync` (`db::commitlog::sync_mode::BATCH`, `use_o_dsync` schema commitlog options).

On enterprise disks (write-through cache) there is no additional fsync due to use_o_dsync:
1. using O_DIRECT means there is nothing in the page cache to sync
2. overwrite means there is no extent metadata to sync
3. with write-through cache, there is no disk cache to sync

On non-enterprise disks, there will be a cost, but writing an sstable is much more expensive.

avikivity commented on schema/schema.hh:
```diff
@@ -573,6 +573,11 @@ struct schema_static_props {
     bool wait_for_sync_to_commitlog = false; // true if all writes using this schema have to be synced immediately by commitlog
     bool use_schema_commitlog = false;
     system_table_load_phase load_phase = system_table_load_phase::phase1;
+    void enable_schema_commitlog() {
```
I feel like this should have a better name, since it implies more things than just the schema commitlog.

If it turns out that we only do this for group0 tables, and we do this for all group0 tables (which is how I think it should be, we don't need this to be extremely fine-tuned), we should just call this setup_for_group0().

Also, we can consider moving it to the schema builder rather than applying it in a separate path.

All of this can be considered for a follow-up.

gusev-p replied:
> If it turns out that we only do this for group0 tables, and we do this for all group0 tables (which is how I think it should be, we don't need this to be extremely fine-tuned), we should just call this setup_for_group0().

It's not group0-specific, now we are turning it on for e.g. `system.local` and `system.scylla_local`

> Also, we can consider moving it to the schema builder rather than applying it in a separate path.

This would probably require adding the bool flag to the 'tables' system table. Now all the `schema_builder` describes should be convertible to mutations. These static configurators [were introduced](https://github.com/scylladb/scylladb/commit/3ef201d67a7521fc82b42e2ba992a16cc337ca5d) to be able attach a piece of config to an arbitrary table, including the one without schema_builder (`CDC_TOPOLOGY_DESCRIPTION`)

avikivity replied:
I agree, it's better to keep this away from the schema.

avikivity commented on main.cc:
```diff
@@ -544,6 +544,85 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
+        cfg.check_experimental(db::experimental_features_t::feature::CONSISTENT_TOPOLOGY_CHANGES);
+}
+
+static void initialize_system_keyspace(sharded<db::system_keyspace>& sys_ks,
```
@xemul should we make scylla_main() a class? though it's not a good fit, the constructor yields like crazy.

gusev-p replied:
> @xemul should we make scylla_main() a class? though it's not a good fit, the constructor yields like crazy.

I think it would be great to have a single class for wiring scylla components together, so that we can reuse it in `scylla_main` and boost tests.

xemul commented on main.cc:
```diff
@@ -545,6 +545,94 @@ static locator::host_id initialize_local_info_thread(sharded<db::system_keyspace
+template <typename... T>
+static void initialize_system_keyspace_thread(sharded<db::system_keyspace>& sys_ks,
+        sharded<cql3::query_processor>& qp,
+        sharded<locator::snitch_ptr>& snitch,
```
Note that `qp` and `snitch` are only needed to call `sys_ks.start()`. If you keep the call to `sharded<system_keyspace>::start()` in main() and make the `initialize_system_keyspace_thread()` be system_keyspace's method (as I suggested) the whole thing would get notably simpler and more self-contained.

Also note, that the latter change would make `db` argument unused too, as `system_keyspace` object already have `database&` reference on board.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3577/

gusev-p commented:
v7:
  * rebased on master, merge conflicts

gusev-p commented:
Should we do something about the check for truncation records in this PR? The one which can produce "has a truncation record. Booting is not safe." error. IIUK, it was introduced to prevent truncation records, which was created based on the old commitlog, to be used in replaying new commitlog. The same problem can happen for any table we switch to the schema commitlog in this PR.

@tgrabiec @kbr-scylla

kbr-scylla commented:
https://github.com/scylladb/scylladb/pull/15279#issuecomment-1708456651

gusev-p commented:
> #15279 (comment)

[No problem](https://github.com/scylladb/scylladb/issues/15354), but the question remains :)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3593/

tgrabiec commented:
> Should we do something about the check for truncation records in this PR? The one which can produce "has a truncation record. Booting is not safe." error. IIUK, it was introduced to prevent truncation records, which was created based on the old commitlog, to be used in replaying new commitlog. The same problem can happen for any table we switch to the schema commitlog in this PR.
> 
> @tgrabiec @kbr-scylla

We should extend the check to be safe. If we would want to support truncation in the future, we can add a commitlog domain selector to truncation records.

xemul commented on gms/feature_service.hh:
```diff
@@ -66,8 +66,11 @@ class feature_service final : public peering_sharded_service<feature_service> {
+    future<> enable_features_on_startup(db::system_keyspace&);
 public:
-    explicit feature_service(feature_config cfg);
+    explicit feature_service(feature_config cfg, bool use_raft_cluster_features);
```
nit: Usually bool_class<> is used for such cases to prevent code reader from wondering what `feature_service f(..., false);` might mean.

avikivity replied:
Or put it into feature_config.

gusev-p replied:
moved it to `feature_config`

kbr-scylla commented on db/schema_tables.cc:
```diff
@@ -1534,13 +1534,14 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
             co_await coroutine::maybe_yield();
         }
+    });
+    co_await db.invoke_on_all([&](replica::database& db) -> future<> {
```
> but maybe there is something that protects us in this specific case

Doubt it. Looks like your assertions catched a bug.
I think an issue would be a good idea.

> Can we ask someone for advice?

Maybe @avikivity ?

avikivity replied:
@tgrabiec

avikivity replied:
@tgrabiec it's "schema_tables: merge_tables_and_views: delay events until tables/view" in case github hides it from you

tgrabiec replied:
Discussed this here: https://github.com/scylladb/scylladb/pull/15279/commits/47ffc66c7f899dbabec66aa096a4827b0f865db2#r1321432290

kbr-scylla commented on replica/table.cc:
```diff
@@ -2040,6 +2041,27 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
+    if (_virtual_writer) [[unlikely]] {
+        return nullptr;
+    }
+    if (_readonly) [[unlikely]] {
```
Hm, instead of two `if`s, I think it would be better to use `if (_readonly && !_virtual_writer)` -- for regular tables in normal operation this will be a single check.

(`return _commitlog` already returns `nullptr` for `_virtual_writer`s, right?)

(I'm worried about minimizing checks here because IIUC `commitlog()` is on the hot path, called for every write (?))

kbr-scylla commented on replica/database.cc:
```diff
@@ -1012,19 +1018,11 @@ future<> database::add_column_family(keyspace& ks, schema_ptr schema, column_fam
 
     if (!readonly) {
-        cf->mark_ready_for_writes();
+        cf->mark_ready_for_writes(commitlog_for(schema));
```
The precondition for calling `commitlog_for` is that `_uses_schema_commitlog` has been set to the correct value -- i.e., we shouldn't call it before `maybe_init_schema_commitlog()`, is that right? So it's incorrect to call `add_column_family(..., false)` before `maybe_init_schema_commitlog()`?

Maybe we should protect ourselves better by making `_uses_schema_commitlog` an `std::optional`, initialize it to `nullopt` and only set it when we make the decision whether we make the decision to use schema commitlog or not after enabling features. In `commitlog_for` we could then verify that it's set and catch bugs.

gusev-p replied:
good point, will update the code, thanks!

gusev-p commented:
v8:
 * inlined `initialize_system_keyspace`;
 * `db.commitlog()`, remove redundant `if`;
 * `use_raft_cluster_features` -> `feature_config`;
 * `db. _uses_schema_commitlog -> optional` to protect against access-before-init bugs;

@kbr-scylla

kbr-scylla commented on db/sstables-format-selector.cc:
```diff
@@ -24,60 +24,75 @@ static const sstring SSTABLE_FORMAT_PARAM_NAME = "sstable_format";
 void feature_enabled_listener::on_enabled() {
     if (!_started) {
         _started = true;
-        _selector.maybe_select_format(_format).get();
```
We should probably have a test to check that sstables created during commitlog replay on startup are created in the right (newest) format.

Can be left to a follow-up.

gusev-p replied:
ok, I'll add it

kbr-scylla reviewed (approved):
Queuing this in ~5 hours if CI passes and no further changes are requested

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3628/

gusev-p commented:
v9:
  * fix boost tests (add `maybe_init_schema_commitlog` into `cql_test_env`)

gusev-p commented:
yet another rebase/merge

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3656/

kbr-scylla commented:
Queued
